public/app/core/context/ModalsContextProvider.tsx

```diff
@@ -0,0 +1,124 @@
+import React, { useEffect, useMemo, useState } from 'react';
+
+import { textUtil } from '@grafana/data';
+import { locationService } from '@grafana/runtime';
+import { ConfirmModal, ConfirmModalProps, ModalsContext } from '@grafana/ui';
+import { ModalsContextState } from '@grafana/ui/src/components/Modal/ModalsContext';
+import { ShowConfirmModalEvent, ShowModalReactEvent } from 'app/types/events';
+
+import appEvents from '../app_events';
+
+export interface Props {
+  children: React.ReactNode;
+}
+
+// eslint-disable-next-line @typescript-eslint/no-explicit-any
+interface StateType<TProps = any> {
```

Comment:
@joshhunt could you go 1 further here and say `<TProps = {}>`? that's how react defines component type:
```
type ComponentType<P = {}> = ComponentClass<P> | FunctionComponent<P>;
```

Reply:
You're right(ish) - that was my first attempt. However, I couldn't convince Typescript that the StateType's TProps is compatible with the showModal's TProps, so I eventually gave up and figured this is good enough.

The more important part to me was the external contract on showModal